OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+    setStateVariableValue(s, "fiber_length", fiberLength);
+    setStateVariableValue(s, "activation",  activation);
+
+}
+// I am cheating here.  Instead of using mli and mdi caches
+/*
+SimTK::Vec2 VandenBogert2011Muscle::getStateVariables(const SimTK::State& s) {
+    double fiberLength = getStateVariableValue(s, "fiber_length");
+    double activ = getStateVariableValue(s, "activation");
+    SimTK::Vec2 stateVariables = {fiberLength, activ};
+    return stateVariables;};
+
+*/
+
+
+//==============================================================================
+// MODELCOMPONENT INTERFACE REQUIREMENTS
+//==============================================================================
+
+void VandenBogert2011Muscle::extendConnectToModel(Model& model)
+{
+    Super::extendConnectToModel(model);
+}
+
+// Define new states and their derivatives in the underlying system
+void VandenBogert2011Muscle::extendAddToSystem(SimTK::MultibodySystem& system)
+const
+{
+
+    Super::extendAddToSystem(system);
+
+    SimTK_ASSERT(isObjectUpToDateWithProperties(),
+                 "VandenBogert2011Muscle: Muscle properties not up-to-date");
+
+    addStateVariable("fiber_length");
+    addStateVariable("activation");
+
+}
+
+void VandenBogert2011Muscle::extendInitStateFromProperties(SimTK::State& s)
+const
+{
+    Super::extendInitStateFromProperties(s);
+
+        setActivation(s, getDefaultActivation());
+        setFiberLength(s, getDefaultFiberLength());
+}
+
+void VandenBogert2011Muscle::extendSetPropertiesFromState(const SimTK::State& s)
+{
+    Super::extendSetPropertiesFromState(s);
+
+        setDefaultActivation(getStateVariableValue(s,"activation"));
+        setDefaultFiberLength(getStateVariableValue(s,"fiber_length"));
+
+}
+
+
+void VandenBogert2011Muscle::
+computeStateVariableDerivatives(const SimTK::State& s) const
+{
+
+    double activ=getActivation(s);
+    // For now we will start with static guess
+    // TODO: figure out how to use previous state as guess
+    SimTK::Vector ydotInitialGuess(2);
+    ydotInitialGuess[0]=0.0;
+    ydotInitialGuess[1]=0.0;
+    SimTK::Vec2 sdot=calcSolveMuscle(s, activ, ydotInitialGuess);
+
+    setStateVariableDerivativeValue(s, "fiber_length", sdot[0]);
+    setStateVariableDerivativeValue(s, "activation", sdot[1]);
+
+
+}
+
+
+//=============================================================================
+// COMPUTATION
+//=============================================================================
+double VandenBogert2011Muscle::fiberLengthToProjectedLength
+        (double fiberLength , bool normalizedValues=0) const {
+
+    double pennAtOptFiberLength = getPennAtOptFiberLength();
+    double projFibLen = 0;
+
+    if (pennAtOptFiberLength < 0.01) {
+        projFibLen = fiberLength;
+    }
+
+    else {
+
+        double optimalFiberLength=getOptimalFiberLength();
+
+        if (normalizedValues) {fiberLength=fiberLength*optimalFiberLength;};
```

Comment:
Thank you for explaining. I think it is fine to have these non-separate functions that handle both the normalized and non-normalized case.

You're right: we don't want to duplicate code. But it may be possible two have separate functions and still not duplicate code (the normalized variant could possibly invoke the non-normalized variant). But we can leave this code as-is until we run into a problem.